### PR TITLE
fix: unecessary program updates by removing timeout methods

### PR DESCRIPTION
**Pull request by sheetalkamat**
This should fix unnecessary program updates that happen right away whenever something changes

### CODE REVIEW

**Comment by bradzacher** [comment]
Thanks for looking into the performance for us! I really appreciate the help 😄 

It looks like this change breaks the persistent parse tests, which ensure that the IDE use case is working as expected.

**Comment by sheetalkamat** [comment]
> cd packages/typescript-estree
yarn test persistentParse

Will do

File: packages/typescript-estree/src/create-program/createWatchProgram.ts [hunk 1]
```diff
@@ -309,25 +309,30 @@ function createWatchProgram(
...
-    if (ms === 250) {
-      cb();
-      return null;
+  watchCompilerHost.trace = log;
+
+  // Since we don't want to asynchronously update program we want to disable timeout methods
+  // So any changes in the program will be delayed and updated when getProgram is called on watch
+  // But because of https://github.com/microsoft/TypeScript/pull/37308 we cannot just set it to undefined
+  // instead save it and call before getProgram is called
+  let callback: (() => void) | undefined;
```
Comment (uniqueiniquity) [inline packages/typescript-estree/src/create-program/createWatchProgram.ts]: So this is introduced for improving pre-TS 3.9 scenarios - does it still make sense when running with https://github.com/microsoft/TypeScript/pull/37308?

Comment (sheetalkamat) [reply packages/typescript-estree/src/create-program/createWatchProgram.ts]: Till we have the release build with that fix, this is the only solution. After that we add condition to just unset these timeout methods for that release onwards .. But both those are essentially almost same// that is do not immediately run those callbacks but wait till someone asks for updated program...

**Comment by sheetalkamat** [comment]
@uniqueiniquity This PR is ready to review and merge with the changes we talked about

**Comment by sheetalkamat** [comment]
And I am not sure failing codecov/patch is about

**Comment by uniqueiniquity** [comment]
@bradzacher any thoughts on the code coverage issue? it's a preemptive change for 3.9 so it makes sense that the path isn't taken.

**Comment by bradzacher** [comment]
If you want to satisfy the branch coverage, you should be able to use istanbul ignore comments to stop the coverage reporting the missed branch.

```ts
/* istanbul ignore if */ if (isRunningNoTimeoutFix) {
// ....
}

/* istanbul ignore else */ if (!isRunningNoTimeoutFix) {
// ....
}
```

Though, I don't treat ever treat code coverage as a hard merge blocker because there's always some edge cases and checks to keep the types "safe".
When it fails I usually review the [coverage report](https://codecov.io/gh/typescript-eslint/typescript-eslint/compare/b5b3be05375c40364d35b887216b56786e33f057...d9a586746a9127b1c160d6d791c993a565603cee/diff#D1-322...324) to see what code was missed, and if it's something that should be tested.

If you're happy with this change, approve it and I'll gladly to merge it.

(again, thanks to you both for looking into all of this for us)

**Comment by sheetalkamat** [comment]
I think we should ignore the coverage result instead of adding ignore comment since as soon as 3.9 releases things should start working...  I will update this with master for merge

**Pull request merged**